README
Irclogger installation
======================
IRC Logger python bot, http://koala.ilog.fr/irclogger

You will need to define 3 places:
   BIN (dir) where to put the CGI web executables and config file
   LOG (dir) where the logs will be stored (ideally out of the web space)
	     and optional html+css style files
   PASS (file) where is the apache passwords file for the channels
and the URL where BIN will be visible from the web
and optionally the CHANNEL to log on startup.

It is developed on linux, should work on all unixes (maybe with the various
utilities such as grep, sed,... in their GNU variants), may work on cygwin.

Quick Install:
==============
uncompress irclogger somewhere (/usr/local/irclogger is the default dir, so 
go to /usr/local, and uncompress the distribution there)
go to this dir
run: ./install BIN LOG PASS URL CHANNEL
follow the instructions.

You need to have agrep installed on your system (http://www.tgries.de/agrep/)
It should work with any 2.x python.

PS: to find quickly where irclogger has been installed on a machine, do
a "locate" to find files irclogger.log (for LOG) and irclogger.conf (for BIN)

Upgrade:
========
Due to the installation using symbolic links, upgrade is a breeze:
  [1] uncompress the distribution where you installed it (default: /usr/local)
  [2] check the HISTORY for other manual upgrade steps, if any
  [3] if the HISTORY says that "irclogger" changed, kill its running 
      instances, you can find them by: 
	   ps awx|grep 'python.*[i]rclogger'
      If you run it via an irclogger-run based script, they will be restarted
      automatically in 45s

Note: major versions (1.4, 1.5...) means that the irclogger python bot has
changed and that you should restart it. Minor versions (1.5a, 1.5b...) means
that only the doc or the cgi scripts have changed, so only step [1] should be
performed. Version of irclogger is specified at the head of irclogger and will
never be minor, version of the rest is in the HISTORY file.

Details:
========

irclogger.conf
--------------
in the BIN directory of the web server
Contains the path of the irclogger main LOG dir (logsdir). Each subdir in it
indicates that a channel of this name is logged.
syntax is bash variable declarations, and can be:

logsdir=/var/log/irclogger
passfile=/var/www/passwds/irclogger-passwds
private_image="${private_image:-<img src='http://koala.ilog.fr/qbullets/private.gif' width=11 height=10 alt='Password-protected log'>}"
top=URL where the upmost "back" link will bring the user

In the irclogger_dir LOG, one can put files:
   style.css  that will be included in all html pages
   message.html will be included at the top of the html index page

LOG must be read&writable by the UID you run the irclogger daemon
and readable by the web server UID
You should do in it:
echo 'deny from all' >.htaccess

Logs are in LOG/CHANNEL/YYYY-MM-DD,Day.log
Logs can be passord-protected by users from the web interface, except
     if a file named "PUBLIC" exists in its dir. Users cannot remove
     the protection (A channel log is protected if a user of the name of the
     channel exists in PASS)

passfile (aka PASS) must be web-server writable htpasswd file, and must exist

Logs can be compressed/decompressed at anytime via gzip. listing and searching
will work transparently with compressed files. If any file is compressed, the
search program will create a expansed mirror of all the logs in LOG/.exp that
should be created writable by the web server UID on installation

to run:
======

copy irclogger-run in /usr/local/bin, edit it, and run it once at startup,
via start-stop-daemon for instance. Copy it under as many copies as you want
to run separate irclogger bots.

On web server:
==============

With a web ScriptAlias dir BIN where you have "AllowOverride All" 
permissions, do:

then:
  S=/usr/local/irclogger
  cd $S
  cp htaccess.sample $BIN/.htaccess
  ln -fs $S/irclogger_* $S/nph-irclogger_reset_auth $BIN

  create a $BIN/irclogger.conf

  cd $BIN
  for i in logs password log log_search 
  do ln -s irclogger_${i} irclogger_${i}_a; done
  # Edit first line of $DIR/.htaccess to reflect $PASS value

  . irclogger.conf; touch $PASS;chmod a+rw $PASS

On your web pages, Call by something like:
<a href='/cgi-bin/irclogger_logs/XXX'>IRC log for channel #XXX</a>

-----------------------------------------------------------------------------

Implementation:
===============

This system is made of scripts running under linux and apache. It should run
on any similar system, such as cygwin on windows systems, but needs the GNU
version of unix tools (date, grep, sed...), not the old SYSV or BSD ones.


irclogger [python]
    Uses ircAsync.py. This script runs the robot
    which connects to channels and dumps logs in text form (MIRC format)
    in $LOG/logs dir
    The rest are shell (bash) scripts.

irclogger-run [bash]
    Is a sample sript to see how to run irclogger. It provides parameters,
    and restarts it on termnination (ircloggers quits in case of
    deconnection)

irclogger_tohtml [bash]
    irclogger_tohtml channel date selected < log
    
    Pretty-prints the channel log. If the channel log dir contains a file named
    "style.css", appends it to the current style (found in the parent dir).
    Possible classes of elements are:
        tr.even  the even rows
        tr.odd   the odd rows
        tr.sep   the separator rows (no messages for more than an hour)
        tr.irc   the control messages
        tr.selected the selection
    selected can be set to a line number to highlight
    If a style.css file is defined, it replaces the one built-in in 
    irclogger_tohtml in var DEFSTYLE

irclogger_totwiki [bash]
    irclogger_totwiki channel date selected < log
    Pretty-prints in a text form incluable in TWiki (see http://twiki.org)

irclogger_totext [bash]
    irclogger_totwiki channel date selected < log
    prints only the contents without control messages and nicknames
    useful for getting meeting notes by a single scribe + correcters

irclogger_logs [bash, cgi]
    irclogger_logs/channel
    lists all available days in channel
    if a file message.html exists in the log dir it will be included at top

irclogger_log [bash, cgi]
    irclogger_log/channel?date=date&raw=on%sel=linenum

irclogger_log_search [bash, cgi]
    irclogger_log_search/channel...

irclogger_log_style [bash, cgi]
    irclogger_log_style channel
    generates a CSS for a channel with a color per active name

nph-irclogger_reset_auth [bash, cgi]
    used to reset the apache authentication to allow login on another
    protected channel

irclogger_password [bash, cgi]
    used to change/set passwords in the PASS file

irclogger_common [bash]
    a library of common functions used by the scripts

style.css [css]
    a sample CSS file to use to have a nicer look


HISTORY
Notes: major versions (1.5, 1.8) indicates that the python IRC bot have been
changed, and should be restarted after upgrade. Minor versions (1.8a, 1.8e) do
not modify the python engine so upgrades are done  by just unpacking the
distribution in place.

1.12a 15 Feb 2005
   MANUAL STEPS for upgrade:
     * from install dir; ln -s $PWD/irclogger_cgiparse $BIN
     (hint: do locate irclogger_log_search to find $BIN dirs)
   fix space not recognised in search on some systems

1.12  25 Jan 2005
   [off] at start of lines can now be case-independent

1.11  30 Sep 2004
    Logged channels status were not saved in case of hard machine crash

1.10a 08 Sep 2004
    new config variable "top" to set topmost back link
    new direct link to search in lchannel in the list of all chans

1.10  02 Aug 2004
    MANUAL STEPS for upgrade:
     * from install dir; ln -s $PWD/{irclogger_logs_cc,irclogger_totext} $BIN
    Minor change to the logger python bot: a debug message was still printed
        when socket was closed.
    install: typo in error message: ccho instead of echo
    irclogger-check: rewrite as a simple invocation of "irc" for more 
        robustness
    direct link to today log (no ?date=) in list of channels
    much faster directory listing with lots of logs (caches in 
        /tmp/.irclogger_logs_cache, settable as conf var cachedir
	(url)/irclogger_logs_cc clears the cache 
    default value for approximate search tolerance is 0 (1 was confusing some)
    new wording for logs list
    new form: contents only, useful for minutes
    in day html view, direct link to next and previous logs

1.9b 19 Mar 2004
    irclogger_log/<CHANNEL> now gives the current day log by a redirection
        so that the URL shown in the browser is perennial

1.9a 06 Jan 2004
    Bugfix in html rendering: 1st line was skipped

1.9  21 Nov 2003
    Change to the logger python bot: to disconnect the bot, the , or : after 
        logger name is now optional: you can say "[LOGGER] off", not just
	"[LOGGER]: off" or "[LOGGER], off"
1.8e 15 Nov 2003
    dates in URLs can now omit the day of week part:
        ...irclogger_log/twiki?date=2003-11-14 is sufficient, no need
	to have ...?date=2003-11-14,Fri
    line ranges can be separated by - only, no more :
1.8d 09 Nov 2003
    selected (highlighted) lines in the html web rendering can now be a range
	of values separated by a dash: -     
	e.g: ?date=2003-11-14,Fri&sel=42-63
1.8c 09 Nov 2003
    log files can now be compressed again.
    you should create a writable by all directory LOG/.exp: do
        for i in `locate irclogger.log`;do mkdir "${i%irclogger.log}.exp"
          chmod a+rw "${i%irclogger.log}.exp";done
1.8b 08 Nov 2003
    search is now done by agrep. As a consequence, logs cannot be gzipped
	anymore if you want to be able to search them (but gzipped files
	are still web-visible and  listable). But agrep has an approximate
	search very useful for IRC logs where typos abound
    you should unzip all you existing logs: go to you LOG dir, and do:
	gunzip */*.log.gz
1.8a 06 Nov 2003
    cgi scripts: uses the $FORM_var system of bashlib rather than the
        `param var` form which seemed to have problems on some machines
1.8 23 Oct 2003
    irclogger: on exit (^C or kill or disconnection) save the list of 
        monitored channels (space-separated list) in file LOG/CHANNELS.last, 
	so that (the new) irclogger-run can reconnect to them on startup.
	No need to restart irclogger if you do not need this feature.
    you will need to update your existing installed irclogger-run scripts with
        the provided patches/irclogger-run-1.8
1.7 23 Oct 2003
    irclogger: forces all channel names to lowercase, ignores previously made
        dirs that were not lowercase. You will have to rename them by hand
	before re-running irclogger.
	This was needed, otherwise if people created channels TWiki and then
	Twiki, the logs were in different dirs although they were the same IRC
	channels.
	No need to restart irclogger if you do not need this feature.
    added a small log converter from xchat format to mirc/irclogger one,
        xchat-to-mirc
1.6a 22 Oct 2003
    html: new output format: TWiki (makes definition lists in TWiki)
        done by new file (cgi bash)  irclogger_totwiki
    Update instructions: make a symbolic link to this file from the BIN dirs:
        in all BINs, do a: ln -s /usr/local/irclogger/irclogger_totwiki .
1.6 21 Oct 2003
    Minor fixs: (python) online bot help gives the URL of the logs
    Niceties to the html page (user manual)
    Decided to go to a major/minor version numbering (no changes to python
        on minor versions)
1.5 20 Oct 2003
    published on the web at http://koala.ilog.fr/irclogger
    entry "irclogger" in freshmeat.net.
1.4 09 Oct 2003
    install script
    irclogger change: log message put in a single line to avoid flooding
         server on startup when initially logging many channels
1.3 04 Oct 2003
    BIN/irclogger.conf variable confdir renamed as logsdir
    (you need to rename the variable in your existing config files when 
    upgrading)
1.2 13 Sep 2003
    should now quit properly on deconnections and not hang indefinitely:
        modified loop code to emit a ping every 60s (tick)
    do not log mess to self even if not the same case
    html: proper per week separation, sunday & saturday smaller
1.1 12 Sep 2003
    Putting a file PUBLIC in a channel log dir makes it unprotectable
1.0 11 Sep 2003
    First fonctional release. 
    Installed on irc.ilog.fr intranet, and on #twiki on overnet
    


ircAsyncD.py
#!/usr/bin/python
"""

ircAsyncD -- An asynchronous IRC client interface.

This is intended as a component in a semantic web agent
with several interfaces, one of them being IRC.
It's implemented on top of asyncore so that the same
agent can export an HTTP interface in asynchronous,
non-blocking style.

see Log at end for recent changes/status info.

Share and Enjoy. Open Source license:
Copyright (c) 2001 W3C (MIT, INRIA, Keio)
http://www.w3.org/Consortium/Legal/copyright-software-19980720

$Id: ircAsyncD.py,v 1.9 2003/09/13 colas Exp $
"""

# asyncore -- Asynchronous socket handler 
# http://www.python.org/doc/current/lib/module-asyncore.html

import string, re
import socket
import asyncore, asynchat




#RFC 2811: Internet Relay Chat: Client Protocol
#2.3 Messages
# http://www.valinor.sorcery.net/docs/rfc2812/2.3-messages.html
SPC="\x20"
CR="\x0d"
LF="\x0a"
CRLF=CR+LF
Port = 6667

# commands...
PRIVMSG = 'PRIVMSG'
NOTICE = 'NOTICE'
PING='PING'
PONG='PONG'
USER='USER'
NICK='NICK'
JOIN='JOIN'
PART='PART'
INVITE='INVITE'
QUIT='QUIT'

# reply codes...
RPL_WELCOME='001'
TOPIC='332'
TOPICINFO='333'

class T(asynchat.async_chat):
    def __init__(self):
        asynchat.async_chat.__init__(self)
        self.bufIn = ''
        self.set_terminator(CRLF)

        # public attributes
        # no blanks in nick.
        # openprojects.net says:
        # Connect with your real username, in lowercase.
        # If your mail address were foo@bar.com, your username would be foo.
        # other limitations? @@see IRC RFC?"""

        self.nick = 'ircAsyncD' # manage this with __getattr__() in stead? hmm...
        self.userid = 'nobody'
        self.fullName = 'ircAsyncD user'
        self._startChannels = ['#test']
        self._dispatch = []
        self._doc = []
        
    def makeConn(self, host, port):
        self.create_socket(socket.AF_INET, socket.SOCK_STREAM)
        debug("connecting to...", host, port)
        self.connect((host, port))

        self.bufIn = ''

    def todo(self, args, *text):
        command = string.join(args)
        if text: command = command + ' :' + string.join(text)

        self.push(command + CRLF)
        debug("sent/pushed command:", command)

    # asyncore methods
    def handle_connect(self):
        debug("connected")

        #@@ hmm... RFC says mode is a bitfield, but
        # irc.py by @@whathisname says +iw string.
        self.todo([NICK, self.nick])
        self.todo([USER, self.userid, "+iw", self.nick], self.fullName)
    def handle_close (self):
        debug('socket closed')
        self.close()

    # asynchat methods
    def collect_incoming_data(self, bytes):
        self.bufIn = self.bufIn + bytes

    def found_terminator(self):
        #debug("found terminator", self.bufIn)
        line = self.bufIn
        self.bufIn = ''

        if line[0] == ':':
            origin, line = string.split(line[1:], ' ', 1)
        else:
            origin = None

        try:
            args, text = string.split(line, ' :', 1)
        except ValueError:
            args = line
            text = ''
        args = string.split(args)

        debug("from::", origin, "|message::", args, "|text::", text)

        self.rxdMsg(args, text, origin)

    def bind(self, thunk, command, textPat=None, doc=None):
        """
        thunk is the routine to bind; it's called ala
          thunk(matchObj or None, origin, args, text)
        
        command is one of the commands above, e.g. PRIVMSG
        textpat is None or a regex object or string to compile to one
        
        doc should be a list of strings; each will go on its own line"""

        if type(textPat) is type(""): textPat = re.compile(textPat)

        self._dispatch.append((command, textPat, thunk))

        if doc: self._doc = self._doc + doc

    def rxdMsg(self, args, text, origin):
        debug('rxdMsg ', args)
        if args[0] == PING:
            self.todo([PONG, text])

        for cmd, pat, thunk in self._dispatch:
            if args[0] == cmd:
                if pat:
                    debug('dispatching on...', pat)
                    m = pat.search(text)
                    if m:
                        thunk(m, origin, args, text)
                else:
                    thunk(None, origin, args, text)

    def startChannels(self, chans):
        self._startChannels = chans
        self.bind(self._welcomeJoin, RPL_WELCOME)
                  
    def _welcomeJoin(self, m, origin, args, text):
        for chan in self._startChannels:
            self.todo(['JOIN', chan])

    def tell(self, dest, text):
        """send a PRIVMSG to dest, a channel or user"""
        self.todo([PRIVMSG, dest], text)

    def notice(self, dest, text):
        """send a NOTICE to dest, a channel or user"""
        self.todo([NOTICE, dest], text)

def actionFmt(str):
    return "\001ACTION" + str + "\001"

def replyTo(myNick, origin, args):
    target = args[1]
    if target == myNick: # just to me
        nick, user, host = splitOrigin(origin)
        return nick
    else:
        return target

def splitOrigin(origin):
    if origin and '!' in origin:
        nick, userHost = string.split(origin, '!', 1)
        if '@' in userHost:
            user, host = string.split(userHost, '@', 1)
        else:
            user, host = userHost, None
    else:
        nick = origin
        user, host = None, None
    return nick, user, host

# cf irc:// urls in Mozilla
# http://www.mozilla.org/projects/rt-messaging/chatzilla/irc-urls.html
# Tue, 20 Mar 2001 21:28:14 GMT

def serverAddr(host, port):
    if port == Port: portPart = ''
    else: portPart = ":%s" % port
    return "irc://%s%s/" % (host, portPart)
        
def chanAddr(host, port, chan):
    if port == Port: portPart = ''
    else: portPart = ":%s" % port
    if chan[0] == '&': chanPart = '%26' + chan[1:]
    elif chan[0] == '#': chanPart = chan[1:]
    else: raise ValueError # dunno what to do with this channel name
    return "irc://%s%s/%s" % (host, portPart, chanPart)
        
def debug(*args):
    import sys
    sys.stderr.write("DEBUG: ")
    for a in args:
        sys.stderr.write(str(a) + ' ')
    sys.stderr.write("\n")


def test(hostName, port, chan):
    c = T()
    c.startChannels([chan])

    def spam(m, origin, args, text, c=c):
        c.tell(args[1], "spam, spam, eggs, and spam!")
    c.bind(spam, PRIVMSG, r"spam\?")

    def bye(m, origin, args, text, c=c):
        c.todo([QUIT], "bye bye!")
    c.bind(bye, PRIVMSG, r"bye bye bot")

    c.makeConn(hostName, port)
    asyncore.loop()
    
    
if __name__=='__main__':
    test('irc.w3.org', 6665, '#rdfbot')
    #test('irc.openprojects.net', Port, '#rdfig')

        
#$Log: ircAsyncD.py,v $
#Revision 1.9 2003/09/13 10:13:28 colas
#removed debug statements, raises EOF on connection close
#
#Revision 1.8  2001/08/24 05:50:52  connolly
#replaced onPrivMsg, onInvite etc. by bind. introduced LOTS of bugs, but I think I fixed most of them. added some documentation, which doubles as test material.
#
#Revision 1.7  2001/08/21 23:03:50  connolly
#general clean-up and tidying... thinking
#about base URIs and contexts.
#
#Revision 1.6  2001/08/21 20:34:12  connolly
#integrated Aaron's patches for join/part.
#holding off on his default prefix patch
#to think about contexts some more.
#
#Revision 1.5  2001/08/20 14:04:56  connolly
#got rid of dead code.
#
#Revision 1.4  2001/08/20 07:39:21  connolly
#questions work now.
#syntax errors are reported.
#irc login protocol revised. (fixed?)
#
#Revision 1.3  2001/08/20 06:16:00  connolly
#woohoo! it's working...
#it accepts RDF/n3 via IRC and stores it in a cwm KB.
#
#Revision 1.2  2001/08/20 04:49:45  connolly
#rdfn3chat has one feature: replies to "help" messages
#with UTSL pointer
#
#Revision 1.1  2001/08/20 03:58:38  connolly
#connects, reponds to PINGs with PONGs.
#


ircAsync.py
#!/usr/bin/python
"""

ircAsync -- An asynchronous IRC client interface.

This is intended as a component in a semantic web agent
with several interfaces, one of them being IRC.
It's implemented on top of asyncore so that the same
agent can export an HTTP interface in asynchronous,
non-blocking style.

see Log at end for recent changes/status info.

Share and Enjoy. Open Source license:
Copyright (c) 2001 W3C (MIT, INRIA, Keio)
http://www.w3.org/Consortium/Legal/copyright-software-19980720

$Id: ircAsync.py,v 1.9 2003/09/13 colas Exp $
"""

# asyncore -- Asynchronous socket handler 
# http://www.python.org/doc/current/lib/module-asyncore.html

import string, re
import socket
import asyncore, asynchat




#RFC 2811: Internet Relay Chat: Client Protocol
#2.3 Messages
# http://www.valinor.sorcery.net/docs/rfc2812/2.3-messages.html
SPC="\x20"
CR="\x0d"
LF="\x0a"
CRLF=CR+LF
Port = 6667

# commands...
PRIVMSG = 'PRIVMSG'
NOTICE = 'NOTICE'
PING='PING'
PONG='PONG'
USER='USER'
NICK='NICK'
JOIN='JOIN'
PART='PART'
INVITE='INVITE'
QUIT='QUIT'

# reply codes...
RPL_WELCOME='001'
TOPIC='332'
TOPICINFO='333'

class T(asynchat.async_chat):
    def __init__(self):
        asynchat.async_chat.__init__(self)
        self.bufIn = ''
        self.set_terminator(CRLF)

        # public attributes
        # no blanks in nick.
        # openprojects.net says:
        # Connect with your real username, in lowercase.
        # If your mail address were foo@bar.com, your username would be foo.
        # other limitations? @@see IRC RFC?"""

        self.nick = 'ircAsync' # manage this with __getattr__() in stead? hmm...
        self.userid = 'nobody'
        self.fullName = 'ircAsync user'
        self._startChannels = ['#test']
        self._dispatch = []
        self._doc = []
        
    def makeConn(self, host, port):
        self.create_socket(socket.AF_INET, socket.SOCK_STREAM)
        #debug("connecting to...", host, port)
        self.connect((host, port))

        self.bufIn = ''

    def todo(self, args, *text):
        command = string.join(args)
        if text: command = command + ' :' + string.join(text)

        self.push(command + CRLF)
        #debug("sent/pushed command:", command)

    # asyncore methods
    def handle_connect(self):
        #debug("connected")

        #@@ hmm... RFC says mode is a bitfield, but
        # irc.py by @@whathisname says +iw string.
        self.todo([NICK, self.nick])
        self.todo([USER, self.userid, "+iw", self.nick], self.fullName)
    def handle_close (self):
        #debug('socket closed')
        self.close()

    # asynchat methods
    def collect_incoming_data(self, bytes):
        self.bufIn = self.bufIn + bytes

    def found_terminator(self):
        #debug("found terminator", self.bufIn)
        line = self.bufIn
        self.bufIn = ''

        if line[0] == ':':
            origin, line = string.split(line[1:], ' ', 1)
        else:
            origin = None

        try:
            args, text = string.split(line, ' :', 1)
        except ValueError:
            args = line
            text = ''
        args = string.split(args)

        #debug("from::", origin, "|message::", args, "|text::", text)

        self.rxdMsg(args, text, origin)

    def bind(self, thunk, command, textPat=None, doc=None):
        """
        thunk is the routine to bind; it's called ala
          thunk(matchObj or None, origin, args, text)
        
        command is one of the commands above, e.g. PRIVMSG
        textpat is None or a regex object or string to compile to one
        
        doc should be a list of strings; each will go on its own line"""

        if type(textPat) is type(""): textPat = re.compile(textPat)

        self._dispatch.append((command, textPat, thunk))

        if doc: self._doc = self._doc + doc

    def rxdMsg(self, args, text, origin):
        if args[0] == PING:
            self.todo([PONG, text])

        for cmd, pat, thunk in self._dispatch:
            if args[0] == cmd:
                if pat:
                    #debug('dispatching on...', pat)
                    m = pat.search(text)
                    if m:
                        thunk(m, origin, args, text)
                else:
                    thunk(None, origin, args, text)

    def startChannels(self, chans):
        self._startChannels = chans
        self.bind(self._welcomeJoin, RPL_WELCOME)
                  
    def _welcomeJoin(self, m, origin, args, text):
        for chan in self._startChannels:
            self.todo(['JOIN', chan])

    def tell(self, dest, text):
        """send a PRIVMSG to dest, a channel or user"""
        self.todo([PRIVMSG, dest], text)

    def notice(self, dest, text):
        """send a NOTICE to dest, a channel or user"""
        self.todo([NOTICE, dest], text)

def actionFmt(str):
    return "\001ACTION" + str + "\001"

def replyTo(myNick, origin, args):
    target = args[1]
    if target == myNick: # just to me
        nick, user, host = splitOrigin(origin)
        return nick
    else:
        return target

def splitOrigin(origin):
    if origin and '!' in origin:
        nick, userHost = string.split(origin, '!', 1)
        if '@' in userHost:
            user, host = string.split(userHost, '@', 1)
        else:
            user, host = userHost, None
    else:
        nick = origin
        user, host = None, None
    return nick, user, host

# cf irc:// urls in Mozilla
# http://www.mozilla.org/projects/rt-messaging/chatzilla/irc-urls.html
# Tue, 20 Mar 2001 21:28:14 GMT

def serverAddr(host, port):
    if port == Port: portPart = ''
    else: portPart = ":%s" % port
    return "irc://%s%s/" % (host, portPart)
        
def chanAddr(host, port, chan):
    if port == Port: portPart = ''
    else: portPart = ":%s" % port
    if chan[0] == '&': chanPart = '%26' + chan[1:]
    elif chan[0] == '#': chanPart = chan[1:]
    else: raise ValueError # dunno what to do with this channel name
    return "irc://%s%s/%s" % (host, portPart, chanPart)
        
def debug(*args):
    import sys
    sys.stderr.write("DEBUG: ")
    for a in args:
        sys.stderr.write(str(a) + ' ')
    sys.stderr.write("\n")


def test(hostName, port, chan):
    c = T()
    c.startChannels([chan])

    def spam(m, origin, args, text, c=c):
        c.tell(args[1], "spam, spam, eggs, and spam!")
    c.bind(spam, PRIVMSG, r"spam\?")

    def bye(m, origin, args, text, c=c):
        c.todo([QUIT], "bye bye!")
    c.bind(bye, PRIVMSG, r"bye bye bot")

    c.makeConn(hostName, port)
    asyncore.loop()
    
    
if __name__=='__main__':
    test('irc.w3.org', 6665, '#rdfbot')
    #test('irc.openprojects.net', Port, '#rdfig')

        
#$Log: ircAsync.py,v $
#Revision 1.9 2003/09/13 10:13:28 colas
#removed debug statements, raises EOF on connection close
#
#Revision 1.8  2001/08/24 05:50:52  connolly
#replaced onPrivMsg, onInvite etc. by bind. introduced LOTS of bugs, but I think I fixed most of them. added some documentation, which doubles as test material.
#
#Revision 1.7  2001/08/21 23:03:50  connolly
#general clean-up and tidying... thinking
#about base URIs and contexts.
#
#Revision 1.6  2001/08/21 20:34:12  connolly
#integrated Aaron's patches for join/part.
#holding off on his default prefix patch
#to think about contexts some more.
#
#Revision 1.5  2001/08/20 14:04:56  connolly
#got rid of dead code.
#
#Revision 1.4  2001/08/20 07:39:21  connolly
#questions work now.
#syntax errors are reported.
#irc login protocol revised. (fixed?)
#
#Revision 1.3  2001/08/20 06:16:00  connolly
#woohoo! it's working...
#it accepts RDF/n3 via IRC and stores it in a cwm KB.
#
#Revision 1.2  2001/08/20 04:49:45  connolly
#rdfn3chat has one feature: replies to "help" messages
#with UTSL pointer
#
#Revision 1.1  2001/08/20 03:58:38  connolly
#connects, reponds to PINGs with PONGs.
#
